Dockable UI

ckaiser opened the pull request:
Here we are! First draft of the finished Dock & Dockable UI

Some things to finish:
- [ ] Take care of all the TODOs.
- [ ] Do more thorough testing with importing older configurations.
- [ ] Improve the logic of how `Dockable`s know where they're docked (aka make the logic).
- [ ] More documentation/comments.
- [ ] Art/theming/icon pass.
- [ ] Testing with popular user themes.

Bonus things:
 - [ ] Tried to auto-hide the `m_actionButton` in the `LayoutSelector::LayoutItem` and then only show it when hovering over the item, but I could NOT get it to work reliably.
 - [ ] Adding the functionality to get widgets to draw themselves onto a surface that we can then draw elsewhere could be useful for one of the TODOs and also maybe other things in future.

dacap committed:
Add Dock widget, initial & basic version of dockable elements (#518)

Some missing features so far:

1) Restore old layout configuration (color bar split pos, timeline
   pos, etc.) and migrate to new Dock layout
2) Load/saving Dock layout
3) Create & customize current layoout (drag-and-drop widgets, etc.)

dacap committed:
Fix memory leaks in MainWindow

Temporary/created subdocks must be deleted automatically, and children
that are not part of the window hierarchy must be deleted explicitly
now (using some std::unique_ptrs).

dacap committed:
Use std::unique_ptr in ToolBar members

dacap committed:
Fix std::clamp() max bound in TipWindow::pointAt()

dacap committed:
Fix popups & tooltips direction when ToolBar is docked at the left side

dacap committed:
Save/restore timeline splitter position correctly

dacap committed:
Save/restore color bar splitter position correctly

dacap committed:
Improve the layout selector UI

Changes:
* Now we use the "user data" icon as the button to expand the layouts
  combobox
* Added a tooltip to this icon
* Added buttons to configure the Timeline position in the same
  combobox
* Fixed some bugs in Dock using space for hidden widgets

dacap committed:
Show the timeline when we set its position from the LayoutSelector

dacap committed:
Set the initial timeline position in the LayoutSelector correctly

dacap committed:
Add option to save/restore user-defined layouts on memory

This happens only in memory at the moment (layouts are not saved in
disk yet), and the customization is quite simple (only size of
splitters, timeline position). But in the future we should be able to
dock elements in any place.

dacap committed:
Save/Load user defined layouts in new user.aseprite-layouts file

And now we store the TiXmlElement for each Layout, instead of
converting from/to text back and forth.

dacap committed:
Include StatusBar in the set of customizable widgets in the layout

dacap committed:
Add possibility to overwrite existent layouts

dacap committed:
Add View > Workspace Layout option to switch the layout w/keys

dacap committed:
Improve borders of context & color bars for both sides (left/right)

dacap committed:
Save layout changes immediately when docks are resized

This patch includes:

* The layout is updated immediately when docks are resized (before
  this the layout was like a snapshot of the configuration when the
  layout was created or overwritten with the "New Layout" option)
* Saving the active layout used in
  preferences.general.workspace_layout so we can restore it after
  restarting Aseprite
* Change "UI Layout" to "Workspace Layout"
* Some strings moved to en.ini file for i18n
* Fixed a crash on MainWindow::onActiveViewChange() when the
  application was being closed

dacap committed:
Update saving/loading dock layouts w/tinyxml2 library

dacap committed:
Fix bug where the ToolBar popup wasn't being opened

This was a new issue with combination of
9b90159d1bdd019e24ef41f6fc76650843d99c2c, probably we should close the
popup when the window is resized (instead of onSizeHint()).

dacap committed:
Show handles in dockable areas to drag-and-drop them

ckaiser committed:
Polished & finalized implementation

dacap commented:
@ckaiser I've just rebased my `dock` branch and cherry-picked the extra commit from `dock-handles` (should be equivalent to what you used as base in 466a6aa86e83da8648af9a04d8517e7803201bf1). You can try rebasing your patch to the new `dock` head and target this PR to the `dock` branch, in this way I can make comments to your patch only.

dacap commented:
There are several direct `TRACE()` calls which would be better to remove them from the patch, or make a conditional `TRACE` (e.g. like in other files [MENU_TRACE](https://github.com/aseprite/aseprite/blob/bbe656c856453b6a3d5cd617f8312417bcf11035/src/app/app_menus.cpp#L46), [CMD_TRACE](https://github.com/aseprite/aseprite/blob/bbe656c856453b6a3d5cd617f8312417bcf11035/src/app/cmd.cpp#L17), etc.)

ckaiser commented:
Rebased and removed some of the superfluous `TRACE()`, only one that might be relevant is when the layout fails to load which I changed to warning log.

dacap commented on src/app/ui/dock.cpp:
```diff
@@ -10,9 +10,19 @@
 #include "app/ui/dockable.h"
 #include "app/ui/skin/skin_theme.h"
+#include "layout_selector.h"
+#include "main_window.h"
```
Please use the full path relative to `src/` folder to include files:
```suggestion
#include "app/ui/layout_selector.h"
#include "app/ui/main_window.h"
```

dacap commented on src/app/ui/dock.cpp:
```diff
@@ -54,34 +64,86 @@ int side_from_index(int index)
+    Paint paint;
+    paint.color(gfx::rgba(0, 0, 0, 200));
+
+    // TODO: This will render any open things, especially the preview editor, need to close or hide
+    // that for a frame or paint the widget itself to a surface instead of croppping the backLayer.
+    auto backLayerSurface = display()->backLayer()->surface();
+    g.drawSurface(backLayerSurface.get(),
+                  dragWidget->bounds(),
+                  gfx::Rect(0, 0, surface->width(), surface->height()),
+                  os::Sampling(),
+                  &paint);
```
Probably we could public the `Widget::paint()` method to paint any widget in any moment on any surface.

dacap commented on src/app/ui/dock.h:
```diff
@@ -21,18 +22,26 @@ namespace app {
-  void onSizeHint(ui::SizeHintEvent& ev) override;
-  void onResize(ui::ResizeEvent& ev) override;
-  void onPaint(ui::PaintEvent& ev) override;
-};
```
I saw that DockTabs was removed, but it was created to implement stacking one dockable widget above another. Anyway if removing this now simplifies the code, we can try to bring it back later to implement the stacking of widgets.

dacap commented on src/app/ui/dock.cpp:
```diff
@@ -422,13 +540,101 @@ bool Dock::onProcessMessage(ui::Message* msg)
+          };
+
+          if (mouseMessage->right() && !m_dragging) {
+            Menu menu;
+            MenuItem left(Strings::dock_left());
+            MenuItem right(Strings::dock_right());
+            MenuItem top(Strings::dock_top());
+            MenuItem bottom(Strings::dock_bottom());
```
Probably we can include the menu item where the widget already is with a check box in this menu.

dacap commented on src/app/ui/dock.cpp:
```diff
@@ -398,22 +440,98 @@ bool Dock::onProcessMessage(ui::Message* msg)
+            break; // Do not handle anything outside the bounds of the dock.
+
+          const int BUFFER_ZONE =
+            std::max(12 * guiscale(), std::min(m_hit.widget->size().w, m_hit.widget->size().h));
```
Probably I'd change this to `kBufferZone`:
```suggestion
          const int kBufferZone =
            std::max(12 * guiscale(), std::min(m_hit.widget->size().w, m_hit.widget->size().h));
```
As all capital letters are reserved more for macros or enums.

dacap commented on src/app/ui/dock.h:
```diff
@@ -8,6 +8,7 @@
 #define APP_UI_DOCK_H_INCLUDED
 #pragma once
 
+#include "dockable.h"
```
Please always use the full path relative to `src/`:
```suggestion
#include "app/ui/dockable.h"
```

dacap commented on src/app/ui/layouts.cpp:
```diff
@@ -25,7 +25,7 @@ using namespace tinyxml2;
 {
   try {
-    std::string fn = m_userLayoutsFilename = UserLayoutsFilename();
+    const std::string& fn = m_userLayoutsFilename = UserLayoutsFilename();
```
I'd try to avoid changing `std::string` to `const std::string&` at the moment (and only doing this if it's 100% necessary for performance issues).

dacap commented on src/app/ui/tabs.cpp:
```diff
@@ -482,6 +483,7 @@ void Tabs::onPaint(PaintEvent& ev)
   auto theme = SkinTheme::get(this);
   Graphics* g = ev.graphics();
   gfx::Rect rect = clientBounds();
+
```
Remove this change.
```suggestion
```

dacap commented on src/app/ui/timeline/timeline.h:
```diff
@@ -155,10 +155,7 @@ class Timeline : public ui::Widget,
-  {
-    return ui::TOP | ui::BOTTOM | ui::LEFT | ui::RIGHT | ui::EXPANSIVE;
-  }
+  int dockableAt() const override { return ui::BOTTOM | ui::LEFT | ui::RIGHT | ui::EXPANSIVE; }
```
AFAIK we can keep the `ui::TOP` side for the `Timeline`, this is not possible in the current `main`/`beta`, but we can add support to move the `Timeline` to the top just keeping the flag.

dacap commented on src/app/ui/dock.cpp:
```diff
@@ -422,13 +540,101 @@ bool Dock::onProcessMessage(ui::Message* msg)
+              if (side & RIGHT) {
+                pos = gen::TimelinePosition::RIGHT;
+              }
+              if (side & BOTTOM) {
+                pos = gen::TimelinePosition::BOTTOM;
+                size.h = (workspaceBounds.h * (1.0 - timelineSplitterPos)) / guiscale();
+              }
```
If side is TOP, we can save the old timeline configuration as bottom, just for backward compatibility.

dacap commented on src/app/ui/layout_selector.cpp:
```diff
@@ -136,88 +99,212 @@ class LayoutSelector::LayoutItem : public ListItem {
-  LayoutPtr m_layout;
+  std::string m_layoutId;
+  std::unique_ptr<IconButton> m_actionButton;
+  obs::connection m_actionConn;
```
Prefer a `obs::scoped_connection` if possible (it automatically disconnects).

dacap reviewed (changes requested):
So far so good, I'm making a first set of review comments so you can make some changes, but I'll require more time (or probably merge it after these fixes and continue from there).

dacap commented on src/app/ui/dock.cpp:
```diff
@@ -422,13 +540,101 @@ bool Dock::onProcessMessage(ui::Message* msg)
+
+          assert(dockableWidget && widgetDock);
+
+          auto dockNRoll = [&](const int side) {
```
About this lambda, I'd prefer a member function if possible, and put another name like `redockWidget()` or something like that.

ckaiser committed:
Codestyle, timeline top support & mark current side in docking menu.

dacap commented:
Merged squashed in https://github.com/dacap/aseprite/commit/c6ae82828d82fd0475d7312cb4e61c12de40bc97 with some notes about the changes in the commit description.
